File: 2022-01-15-jpa-one-plus-n-problem/action-in-blog/src/test/java/blog/in/action/reply/ReplyRepositoryTest.java
package blog.in.action.reply;

import blog.in.action.post.Post;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;

import javax.persistence.EntityManager;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

@DataJpaTest
public class ReplyRepositoryTest {

    @Autowired
    private EntityManager em;

    @Autowired
    private ReplyRepository replyRepository;

    @BeforeEach
    public void setup() {

        Post post = Post.builder()
                .title("first post")
                .content("this is the first post.")
                .build();

        em.persist(post);

        for (int index = 0; index < 10; index++) {
            Reply reply = Reply.builder()
                    .content("reply-" + index)
                    .post(post)
                    .build();
            post.addReply(reply);

            replyRepository.save(reply);
        }

        em.flush();
        em.clear();
    }

    @Test
    public void givenFindByContent_whenGetPost_thenNPlusOneQuery() {

        Reply reply = replyRepository.findByContent("reply-0");

        Post post = reply.getPost();

        assertThat(post.getTitle()).isEqualTo("first post");
    }

    @Test
    public void whenFindByContentFetchJoin_thenJustOneQuery() {

        Reply reply = replyRepository.findByContentFetchJoin("reply-0");

        Post post = reply.getPost();

        assertThat(post.getTitle()).isEqualTo("first post");
    }

    @Test
    public void whenFindByContentEntityGraph_thenJustOneQuery() {

        Reply reply = replyRepository.findByContentEntityGraph("reply-0");

        Post post = reply.getPost();

        assertThat(post.getTitle()).isEqualTo("first post");
    }
}

File: 2024-03-21-problem-of-test-container-inside-super-class/action-in-blog/src/test/kotlin/blog/in/action/repository/ReplyRepositoryTest.kt
package blog.`in`.action.repository

import blog.`in`.action.TestStoreConfig
import blog.`in`.action.domain.ReplyEntity
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired

class ReplyRepositoryTest : TestStoreConfig() {

    @Autowired
    lateinit var sut: ReplyRepository

    @Test
    fun findByContentContains() {
        val entity = ReplyEntity(content = "This is new hello world")
        entityManager.persist(entity)
        flushAndClear()


        val result = sut.findByContentContains("hello world")


        assertEquals(1, result.size)
        assertEquals("This is new hello world", result[0].content)
    }
}